openlibrary/plugins/admin/code.py

```diff
@@ -239,6 +239,85 @@ def GET(self):
         return delegate.RawText(json.dumps(data), content_type="application/json")
 
 
+class sync_ia_ol(delegate.page):
+    path = '/ia/sync'
+    encoding = 'json'
+
+    def POST(self):
+        # Authenticate request:
+        s3_access_key = web.ctx.env.get('HTTP_X_S3_ACCESS', '')
+        s3_secret_key = web.ctx.env.get('HTTP_X_S3_SECRET', '')
+
+        if not self.is_authorized(s3_access_key, s3_secret_key):
+            raise web.unauthorized()
+
+        # Validate input
+        i = json.loads(web.data())
+
+        if not self.validate_input(i):
+            raise web.badrequest('Missing required fields')
+
+        # Find record using OLID (raise 404 if not found)
+        edition_key = f'/books/{i.get("olid")}'
+        edition = web.ctx.site.get(edition_key)
+
+        if not edition:
+            raise web.notfound()
+
+        # Update record
+        match i.get('action', ''):
+            case 'remove':
+                self.remove_ocaid(edition)
+            case 'modify':
+                self.modify_ocaid(edition, i.get('ocaid'))
+            case '_':
+                raise web.badrequest('Unknown action')
+
+        return delegate.RawText(json.dumps({"status": "ok"}))
+
+    @cache.memoize(
+        engine="memcache",
+        key=lambda self, access, secret: f'ia_sync-{access}',
+        expires=5 * 60,
+    )
```

Comment:
```suggestion
```
With this type of caching in place, a malicious person could modify `ocaid`s with only the S3 access key (if a genuine request was live in cache).